File: packages/retrieval/eval/longmemeval/embed.ts
import { createHash } from 'node:crypto';
import { mkdir, readFile, writeFile } from 'node:fs/promises';
import { dirname } from 'node:path';
import type { Embedder } from './types';

const MOCK_DIM = 256;
const OPENAI_MODEL = 'text-embedding-3-small';
const OPENAI_DIM = 1536;

function tokenize(text: string): string[] {
  return text
    .toLowerCase()
    .split(/[^a-z0-9]+/)
    .filter((t) => t.length > 0);
}

function l2normalize(vec: number[]): number[] {
  const norm = Math.sqrt(vec.reduce((sum, v) => sum + v * v, 0)) || 1;
  return vec.map((v) => v / norm);
}

/**
 * Deterministic hashed bag-of-words embedding. Each token is hashed into a few
 * fixed dimensions; lexical overlap raises cosine similarity. Enough to prove
 * ranking, not a real semantic score. No network, no keys.
 */
export function createMockEmbedder(dim = MOCK_DIM): Embedder {
  return {
    name: `mock-hashed-bow(dim=${dim})`,
    dims: dim,
    embed: async (text: string) => {
      const vec = new Array<number>(dim).fill(0);
      for (const token of tokenize(text)) {
        const h = createHash('sha256').update(token).digest();
        // Spread each token across 4 slots with signed weights.
        for (let s = 0; s < 4; s++) {
          const idx = h.readUInt32LE(s * 4) % dim;
          const sign = (h[s * 4 + 3] & 1) === 0 ? 1 : -1;
          vec[idx] += sign;
        }
      }
      return l2normalize(vec);
    },
  };
}

interface EmbedCache {
  get(key: string): number[] | undefined;
  set(key: string, vec: number[]): void;
  flush(): Promise<void>;
}

async function loadCache(path: string): Promise<EmbedCache> {
  let map = new Map<string, number[]>();
  let dirty = false;
  try {
    const raw = await readFile(path, 'utf8');
    map = new Map(Object.entries(JSON.parse(raw) as Record<string, number[]>));
  } catch {
    // No cache yet; start empty.
  }
  return {
    get: (key) => map.get(key),
    set: (key, vec) => {
      map.set(key, vec);
      dirty = true;
    },
    flush: async () => {
      if (!dirty) return;
      try {
        await mkdir(dirname(path), { recursive: true });
        await writeFile(path, JSON.stringify(Object.fromEntries(map)));
        dirty = false;
      } catch (err) {
        // The on-disk cache is only a cost optimization. At large scale the map
        // can exceed V8's max string length when serialized; never let a flush
        // failure discard an otherwise-completed eval — warn and continue.
        console.warn(`embedding cache flush skipped: ${(err as Error).message}`);
      }
    },
  };
}

function cacheKey(text: string): string {
  return createHash('sha256').update(`${OPENAI_MODEL}\n${text}`).digest('hex');
}

// OpenAI's embeddings endpoint accepts up to 2048 inputs and ~300k tokens per
// request. Batch by both a count cap and a cumulative-char budget (~4 chars/
// token heuristic → ~150k tokens) so a batch never trips the token limit.
const MAX_BATCH_INPUTS = 256;
const MAX_BATCH_CHARS = 600_000;

const sleep = (ms: number): Promise<void> => new Promise((resolve) => setTimeout(resolve, ms));

// Batching is fast enough to exceed OpenAI's tokens-per-minute (TPM) cap, which
// returns 429. Retry on 429 (honoring Retry-After / the suggested delay) and on
// transient 5xx with exponential backoff, so the embedder paces itself to the
// TPM limit instead of aborting a long run.
const MAX_RETRIES = 8;

function retryDelayMs(res: Response, body: string, attempt: number): number {
  const header = res.headers.get('retry-after');
  if (header !== null) {
    const secs = Number(header);
    if (Number.isFinite(secs)) return secs * 1000;
  }
  const match = body.match(/try again in ([\d.]+)(ms|s)/i);
  if (match) {
    const value = Number(match[1]);
    return match[2].toLowerCase() === 's' ? value * 1000 : value;
  }
  return Math.min(2 ** attempt * 500, 20_000);
}

async function embedInputs(apiKey: string, inputs: string[]): Promise<number[][]> {
  for (let attempt = 0; ; attempt++) {
    const res = await fetch('https://api.openai.com/v1/embeddings', {
      method: 'POST',
      headers: {
        'Content-Type': 'application/json',
        Authorization: `Bearer ${apiKey}`,
      },
      body: JSON.stringify({ model: OPENAI_MODEL, input: inputs }),
    });
    if (res.ok) {
      const json = (await res.json()) as { data: { index: number; embedding: number[] }[] };
      // Order is not guaranteed; place each embedding at its reported index.
      const out = new Array<number[]>(inputs.length);
      for (const row of json.data) out[row.index] = row.embedding;
      return out;
    }
    const body = await res.text();
    // An input exceeds the model's token limit. The chunker bounds chunks by a
    // local token estimate, but OpenAI's tokenizer can count slightly higher, so
    // self-heal instead of aborting a long, expensive run: halve the batch to
    // isolate the offending input, then halve a lone over-long input until it
    // fits. Order is preserved so results still align with `inputs`.
    if (res.status === 400 && /maximum input length/i.test(body)) {
      if (inputs.length > 1) {
        const mid = Math.ceil(inputs.length / 2);
        const head = await embedInputs(apiKey, inputs.slice(0, mid));
        const tail = await embedInputs(apiKey, inputs.slice(mid));
        return [...head, ...tail];
      }
      if (inputs[0].length > 1) {
        const half = inputs[0].slice(0, Math.floor(inputs[0].length / 2));
        return embedInputs(apiKey, [half]);
      }
    }
    const retryable = res.status === 429 || res.status >= 500;
    if (!retryable || attempt >= MAX_RETRIES) {
      throw new Error(`OpenAI embeddings failed (${res.status}): ${body.slice(0, 300)}`);
    }
    // Add a little headroom over the suggested delay so the next request lands
    // after the rate window has actually rolled over.
    await sleep(retryDelayMs(res, body, attempt) + 250);
  }
}

/**
 * Real OpenAI text-embedding-3-small (1536 dims) with an on-disk,
 * content-addressed cache so re-runs are cheap and indexing isn't re-billed.
 */
export async function createOpenAIEmbedder(
  apiKey: string,
  cachePath: string,
): Promise<Embedder> {
  const cache = await loadCache(cachePath);
  return {
    name: `openai:${OPENAI_MODEL}`,
    dims: OPENAI_DIM,
    embed: async (text: string) => {
      const key = cacheKey(text);
      const cached = cache.get(key);
      if (cached !== undefined) return cached;
      const [vec] = await embedInputs(apiKey, [text]);
      cache.set(key, vec);
      return vec;
    },
    prewarm: async (texts: string[]) => {
      // Embed only texts not already cached, in token-bounded batches, so the
      // per-entry `embed` calls during upsert all become cache hits.
      const pending = [...new Set(texts)].filter((t) => cache.get(cacheKey(t)) === undefined);
      let batch: string[] = [];
      let chars = 0;
      const drain = async (): Promise<void> => {
        if (batch.length === 0) return;
        const vecs = await embedInputs(apiKey, batch);
        batch.forEach((t, i) => cache.set(cacheKey(t), vecs[i]));
        batch = [];
        chars = 0;
      };
      for (const text of pending) {
        if (batch.length >= MAX_BATCH_INPUTS || chars + text.length > MAX_BATCH_CHARS) {
          await drain();
        }
        batch.push(text);
        chars += text.length;
      }
      await drain();
    },
    flush: () => cache.flush(),
  };
}

File: packages/retrieval-py/eval/longmemeval/embed.py
from __future__ import annotations

import hashlib
import json
import math
import re
from pathlib import Path

from .openai_http import post_json
from .types import Embedder

MOCK_DIM = 256
OPENAI_MODEL = "text-embedding-3-small"
OPENAI_DIM = 1536
_OPENAI_URL = "https://api.openai.com/v1/embeddings"

_TOKEN_RE = re.compile(r"[^a-z0-9]+")


def _tokenize(text: str) -> list[str]:
    return [t for t in _TOKEN_RE.split(text.lower()) if len(t) > 0]


def _l2normalize(vec: list[float]) -> list[float]:
    norm = math.sqrt(sum(v * v for v in vec)) or 1.0
    return [v / norm for v in vec]


class _MockEmbedder:
    """Deterministic hashed bag-of-words embedding. Each token is hashed into a
    few fixed dimensions; lexical overlap raises cosine similarity. Enough to
    prove ranking, not a real semantic score. No network, no keys.
    """

    def __init__(self, dim: int = MOCK_DIM) -> None:
        self.name = f"mock-hashed-bow(dim={dim})"
        self.dims = dim
        self._dim = dim

    async def embed(self, text: str) -> list[float]:
        vec = [0.0] * self._dim
        for token in _tokenize(text):
            h = hashlib.sha256(token.encode("utf8")).digest()
            # Spread each token across 4 slots with signed weights.
            for s in range(4):
                idx = int.from_bytes(h[s * 4 : s * 4 + 4], "little") % self._dim
                sign = 1 if (h[s * 4 + 3] & 1) == 0 else -1
                vec[idx] += sign
        return _l2normalize(vec)

    async def flush(self) -> None:
        return None


def create_mock_embedder(dim: int = MOCK_DIM) -> Embedder:
    return _MockEmbedder(dim)


class _EmbedCache:
    def __init__(self, path: str) -> None:
        self._path = path
        self._dirty = False
        self._map: dict[str, list[float]] = {}
        try:
            loaded = json.loads(Path(path).read_text(encoding="utf8"))
        except (OSError, ValueError):
            # No cache yet; start empty.
            loaded = {}
        # A valid-but-non-object cache (null, a list, ...) must not replace the
        # dict, or the first get()/set() would raise AttributeError.
        if isinstance(loaded, dict):
            self._map = loaded

    def get(self, key: str) -> list[float] | None:
        return self._map.get(key)

    def set(self, key: str, vec: list[float]) -> None:
        self._map[key] = vec
        self._dirty = True

    def flush(self) -> None:
        if not self._dirty:
            return
        try:
            path = Path(self._path)
            path.parent.mkdir(parents=True, exist_ok=True)
            path.write_text(json.dumps(self._map), encoding="utf8")
            self._dirty = False
        except OSError as err:
            # The on-disk cache is only a cost optimization; never let a flush
            # failure discard an otherwise-completed eval — warn and continue.
            print(f"embedding cache flush skipped: {err}")


class _OpenAIEmbedder:
    """Real OpenAI text-embedding-3-small (1536 dims) with an on-disk,
    content-addressed cache so re-runs are cheap and indexing isn't re-billed.
    """

    def __init__(self, api_key: str, cache_path: str) -> None:
        self.name = f"openai:{OPENAI_MODEL}"
        self.dims = OPENAI_DIM
        self._api_key = api_key
        self._cache = _EmbedCache(cache_path)

    async def embed(self, text: str) -> list[float]:
        key = hashlib.sha256(f"{OPENAI_MODEL}\n{text}".encode("utf8")).hexdigest()
        cached = self._cache.get(key)
        if cached is not None:
            return cached
        json_resp = await post_json(
            _OPENAI_URL, self._api_key, {"model": OPENAI_MODEL, "input": text}, "embeddings"
        )
        vec = json_resp["data"][0]["embedding"]
        self._cache.set(key, vec)
        return vec

    async def flush(self) -> None:
        self._cache.flush()


def create_openai_embedder(api_key: str, cache_path: str) -> Embedder:
    return _OpenAIEmbedder(api_key, cache_path)
